components/homme/src/share/gllfvremap_mod.F90

```diff
@@ -0,0 +1,2297 @@
+    integer, intent(in) :: nt
+    real(kind=real_kind), intent(in) :: dt
+    type (hvcoord_t), intent(in) :: hvcoord
+    type (element_t), intent(inout) :: elem(:)
+
+    type (hybrid_t) :: hybrid
+    integer :: nets, nete
+    
+#ifdef HORIZ_OPENMP
+    !$omp parallel num_threads(hthreads), default(shared), private(nets,nete,hybrid)
+#endif
+    call gfr_hybrid_create(par, dom_mt, hybrid, nets, nete)
+    call gfr_pg1_reconstruct_hybrid(hybrid, nt, dt, hvcoord, elem, nets, nete)
+#ifdef HORIZ_OPENMP
+    !$omp end parallel
+#endif
+  end subroutine gfr_pg1_reconstruct_dom_mt
+
+  subroutine gfr_pg1_reconstruct_topo_dom_mt(par, dom_mt, elem)
+    ! Wrapper to the hybrid-threading main routine.
+
+    use parallel_mod, only: parallel_t
+    use domain_mod, only: domain1d_t
+    use thread_mod, only: hthreads
+
+    type (parallel_t), intent(in) :: par
+    type (domain1d_t), intent(in) :: dom_mt(:)
+    type (element_t), intent(inout) :: elem(:)
+
+    type (hybrid_t) :: hybrid
+    integer :: nets, nete
+    
+#ifdef HORIZ_OPENMP
+    !$omp parallel num_threads(hthreads), default(shared), private(nets,nete,hybrid)
+#endif
+    call gfr_hybrid_create(par, dom_mt, hybrid, nets, nete)
+    call gfr_pg1_reconstruct_topo_hybrid(hybrid, elem, nets, nete)
+#ifdef HORIZ_OPENMP
+    !$omp end parallel
+#endif
+  end subroutine gfr_pg1_reconstruct_topo_dom_mt
+
+  ! ----------------------------------------------------------------------
+  ! Internal helpers.
+
+  subroutine gfr_dyn_to_fv_phys_topo_elem(elem, ie, phis)
+    ! Remap topo within an element.
+
+    type (element_t), intent(in) :: elem(:)
+    integer, intent(in) :: ie
+    real(kind=real_kind), intent(out) :: phis(:)
+
+    real(kind=real_kind) :: wr(np,np,2), ones(np,np), qmin, qmax
+    integer :: nf, nf2, ncol
+
+    ones = one
+    nf = gfr%nphys
+    nf2 = nf*nf
+    ncol = nf*nf
+
+    wr(:,:,1) = elem(ie)%state%phis(:,:)
+    call gfr_g2f_scalar(ie, elem(ie)%metdet, wr(:,:,1:1), wr(:,:,2:2))
+    qmin = minval(elem(ie)%state%phis)
+    qmax = maxval(elem(ie)%state%phis)
+    wr(:nf,:nf,1) = reshape(gfr%w_ff(:nf2)*gfr%fv_metdet(:nf2,ie), (/nf,nf/))
+    call limiter_clip_and_sum(gfr%nphys, wr(:,:,1), qmin, qmax, ones, wr(:nf,:nf,2))
+    phis(:ncol) = reshape(wr(:nf,:nf,2), (/ncol/))
+  end subroutine gfr_dyn_to_fv_phys_topo_elem
+
+  subroutine gfr_hybrid_create(par, dom_mt, hybrid, nets, nete)
+    ! Create a hybrid_t given a dom_mt.
+
+    use parallel_mod, only: parallel_t
+    use domain_mod, only: domain1d_t
+    use hybrid_mod, only: hybrid_t, hybrid_create
+    use thread_mod, only: omp_get_thread_num, hthreads
+
+    type (parallel_t), intent(in) :: par
+    type (domain1d_t), intent(in) :: dom_mt(:)
+    type (hybrid_t), intent(out) :: hybrid
+    integer, intent(out) :: nets, nete
+
+    integer :: ithr
+
+    ithr = omp_get_thread_num()
+    nets = dom_mt(ithr+1)%start
+    nete = dom_mt(ithr+1)%end
+    hybrid = hybrid_create(par, ithr, hthreads)
```

Comment:
Should the access to `dom_mt` be with `ithr` and not `ithr+1` for zero-length arrays? A seg-fault shows up with `SMS_Ln5.ne4pg2_ne4pg2.FC5AV1C-L.anvil_intel.cam-thetahy_pg2` (MPI-only run with 108 tasks for ATM) for tasks 96 and up: e.g.
```
106: forrtl: severe (174): SIGSEGV, segmentation fault occurred
106: Image              PC                Routine            Line        Source
106: e3sm.exe           0000000002981B51  Unknown               Unknown  Unknown
106: e3sm.exe           000000000297FC8B  Unknown               Unknown  Unknown
106: e3sm.exe           000000000290F104  Unknown               Unknown  Unknown
106: e3sm.exe           000000000290EF16  Unknown               Unknown  Unknown
106: e3sm.exe           000000000288B147  Unknown               Unknown  Unknown
106: e3sm.exe           0000000002896D40  Unknown               Unknown  Unknown
106: libpthread-2.12.s  00002B87996807E0  Unknown               Unknown  Unknown
106: e3sm.exe           000000000127A829  gllfvremap_mod_mp        1769  gllfvremap_mod.F90
106: e3sm.exe           00000000016897FA  inidat_mp_read_in         466  inidat.F90
106: e3sm.exe           00000000013CF77D  startup_initialco          54  startup_initialconds.F90
106: e3sm.exe           00000000011EB75F  inital_mp_cam_ini          42  inital.F90
106: e3sm.exe           0000000000500F40  cam_comp_mp_cam_i         159  cam_comp.F90
106: e3sm.exe           00000000004F7E39  atm_comp_mct_mp_a         312  atm_comp_mct.F90
106: e3sm.exe           000000000043B808  component_mod_mp_         257  component_mod.F90
106: e3sm.exe           000000000042A861  cime_comp_mod_mp_        1337  cime_comp_mod.F90
106: e3sm.exe           00000000004388B9  MAIN__                    122  cime_driver.F90
```

Reply:
It should definitely be (ithr+1). I actually ran into a very similar issue, but I'm having trouble remembering what it was. I'm gonna look back through some slack conversations with Andrew to try and recall the details.

Reply:
No. There are two separate issues here. I will need to fix the second.

The first is whether `ithr` or `ithr+1` should be used for indexing in general, given that `ithr` starts at 0. This depends on how `dom_mt` is declared. If it's declared as a `pointer`, then it inherits the index space from the allocation of the pointer. If it's declared as an assumed-shape array, then it does not inherit the index space. This issue comes up enough that I've created a single-file reproducer of the various cases:
```F90
module ptr
contains
  subroutine ptr1(a)
    real, pointer, intent(in) :: a(:) ! pointer attribute means a has index space of allocation
    print *,'ptr1',a(0)
  end subroutine ptr1

  subroutine ptr2(a)
    real, intent(in) :: a(:) ! non-pointer attribute means a has default index space
    print *,'ptr2',a(1)
  end subroutine ptr2

  subroutine ptr3(a)
    real, intent(in) :: a(4:6) ! non-pointer attribute also allows reindexing
    print *,'ptr3',a(4)
  end subroutine ptr3

#if 0
  subroutine ptr4(a)
    real, pointer, intent(in) :: a(4:6) ! not allowed
    print *,'ptr4',a(4)
  end subroutine ptr4
#endif
end module ptr

program main
  use ptr

  real, pointer :: a(:)

  allocate(a(0:3))
  a(0) = 42
  a(1) = 2
  call ptr1(a)
  call ptr2(a)
  call ptr3(a)
  deallocate(a)
end program main
```
Thus, `ithr+1` is correct in this case.

However, the second issue is a genuine bug. Here, for `ne=4`, there are 96 elements. Thus, if more than 96 procs are used, then this code is indexing a dom_mt that has 0 length. No index value is correct in this case. I need to fix this code so if an MPI rank has 0 elements, it doesn't do anything.

Reply:
I've opened https://github.com/E3SM-Project/E3SM/issues/3241 and will create a PR to fix this in the next few days.

Reply:
Just curious, wouldn't this declaration work with `dom_mt(ithr)` access?
```
type (domain1d_t), pointer, intent(in) :: dom_mt(:)
```

Reply:
To be clear, the issue you found relates to having more processes than elements. Declaration of dom_mt is orthogonal to that problem.

Re: declaring dom_mt: As a matter of personal taste, I don't like to declare something as a pointer unless it needs to be, as doing so narrows the scope of application of the subroutine. In this case, taking an assumed-shape array as input provides the greatest generality.

Reply:
Yes, that works, but another seg-fault occurs at `hybrid_create`: need the `par%dynproc` check.

Reply:
Right, exactly. Unless I'm mistaken, PR #3242 takes care of exactly the issue of including a dynproc check at all necessary spots. As I report in that PR, the 108-proc run for the 96-element case now completes. Thanks again for reporting the issue.